src/shared/conf-parser.c

```diff
@@ -1398,21 +1398,48 @@ int config_parse_compression(
                 void *data,
                 void *userdata) {
 
-        Compression *compression = ASSERT_PTR(data);
+        CompressionOpts ***opts = ASSERT_PTR(data);
+        int level = -1;
+        char *l = NULL;
+        _cleanup_free_ char *t = NULL;
 
         if (isempty(rvalue)) {
-                *compression = COMPRESSION_NONE;
                 return 1;
         }
 
-        Compression c = compression_lowercase_from_string(rvalue);
+        l = strchr(rvalue, ':');
+        if (l) {
+                double lVal = strtol(l, NULL, 10);
+                if (lVal < 0)
+                        return log_syntax(unit, LOG_WARNING, filename, line, 0,
+                                          "Compression level %s should be positive", l);
+                level = lVal;
+        }
+
+        t = strndup(rvalue, l - rvalue);
+        if (!t)
+                return log_syntax_parse_error(unit, filename, line, ENOMEM, lvalue, rvalue);
+
+        Compression c = compression_lowercase_from_string(t);
         if (c < 0)
                 return log_syntax_parse_error(unit, filename, line, c, lvalue, rvalue);
         if (!compression_supported(c))
                 return log_syntax(unit, LOG_WARNING, filename, line, 0,
-                                  "Compression=%s is not supported on a system", rvalue);
+                                  "Compression=%s is not supported on a system", t);
 
-        *compression = c;
+        int size = 0;
+        for (int i = 0; *opts != NULL && (*opts)[i] != NULL; i++) {
+                if ((*opts)[i]->algorithm == c) {
+                        (*opts)[i]->level = level;
+                        return 1;
+                }
+                size++;
+        }
+        *opts = (CompressionOpts**)realloc(*opts, (size + 2) * sizeof(CompressionOpts*));
+        (*opts)[size] = (CompressionOpts*)malloc(sizeof(CompressionOpts));
+        (*opts)[size]->level = level;
+        (*opts)[size]->algorithm = c;
+        (*opts)[size+1] = NULL;
```

Comment:
instead of having an array of pointers with NULL termination i think we shuld have an array directly, and accompany it with a size param.